source/iterator/iterator.go

```diff
@@ -0,0 +1,303 @@
+	columnOperationType = "CONDUIT_OPERATION_TYPE"
+	columnTimeCreated   = "CONDUIT_TRACKING_CREATED_DATE"
+	columnTrackingID    = "CONDUIT_TRACKING_ID"
+)
+
+// CombinedIterator combined iterator.
+type CombinedIterator struct {
+	cdc      *CDCIterator
+	snapshot *SnapshotIterator
+
+	// connection string.
+	conn string
+
+	// table - table name.
+	table string
+	// trackingTable - tracking table name.
+	trackingTable string
+	// columns list of table columns for record payload
+	// if empty - will get all columns.
+	columns []string
+	// key Name of column what iterator use for setting key in record.
+	key string
+	// orderingColumn Name of column what iterator use for sorting data.
+	orderingColumn string
+	// batchSize size of batch.
+	batchSize int
+	// columnTypes column types from table.
+	columnTypes map[string]string
+}
+
+// NewCombinedIterator - create new iterator.
+func NewCombinedIterator(
+	ctx context.Context,
+	db *sqlx.DB,
+	conn, table, key, orderingColumn string,
+	columns []string,
+	batchSize int,
+	sdkPosition sdk.Position,
+) (*CombinedIterator, error) {
+	var err error
+
+	it := &CombinedIterator{
+		conn:           conn,
+		table:          table,
+		columns:        columns,
+		key:            key,
+		orderingColumn: orderingColumn,
+		batchSize:      batchSize,
+		trackingTable:  fmt.Sprintf(trackingTablePattern, table),
+	}
+
+	// get column types for converting.
+	it.columnTypes, err = coltypes.GetColumnTypes(ctx, db, table)
+	if err != nil {
+		return nil, fmt.Errorf("get table column types: %w", err)
+	}
+
+	// create tracking table, create triggers for cdc logic.
+	err = it.SetupCDC(ctx, db)
+	if err != nil {
+		return nil, fmt.Errorf("setup cdc: %w", err)
+	}
+
+	pos, err := position.ParseSDKPosition(sdkPosition)
+	if err != nil {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	if pos == nil || pos.IteratorType == position.TypeSnapshot {
+		it.snapshot, err = NewSnapshotIterator(ctx, db, table, orderingColumn, key, columns,
+			batchSize, pos, it.columnTypes)
+		if err != nil {
+			return nil, fmt.Errorf("new shapshot iterator: %w", err)
+		}
+	} else {
+		it.cdc, err = NewCDCIterator(ctx, db, it.table, it.trackingTable, it.key,
+			it.columns, it.batchSize, pos)
+		if err != nil {
+			return nil, fmt.Errorf("new shapshot iterator: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// SetupCDC - create tracking table, add columns, add triggers, set identity column.
+func (c *CombinedIterator) SetupCDC(ctx context.Context, db *sqlx.DB) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return fmt.Errorf("create transaction: %w", err)
+	}
+
+	defer tx.Rollback() // nolint:errcheck,nolintlint
+
+	// check if table exist.
+	rows, err := tx.QueryContext(ctx, fmt.Sprintf(queryIfExistTable, c.trackingTable))
+	if err != nil {
+		return fmt.Errorf("query exist table: %w", err)
+	}
+
+	defer rows.Close() //nolint:staticcheck,nolintlint
+
+	for rows.Next() {
+		var count int
+		er := rows.Scan(&count)
```

Comment:
Why not use `err` instead of `er`?

Reply:
Because on this case we shadow variable err